--- dehydrated.orig	2019-03-27 16:08:47.663360357 +0100
+++ dehydrated	2019-03-28 16:22:23.939282353 +0100
@@ -94,7 +94,7 @@
 
 # verify configuration values
 verify_config() {
-  [[ "${CHALLENGETYPE}" == "http-01" || "${CHALLENGETYPE}" == "dns-01" || "${CHALLENGETYPE}" == "tls-alpn-01" ]] || _exiterr "Unknown challenge type ${CHALLENGETYPE}... cannot continue."
+  [[ "${CHALLENGETYPE}" == "http-01" || "${CHALLENGETYPE}" == "dns-01" || "${CHALLENGETYPE}" == "tls-alpn-01" || "${CHALLENGETYPE}" == "x-sunet-01" ]] || _exiterr "Unknown challenge type ${CHALLENGETYPE}... cannot continue."
   if [[ "${CHALLENGETYPE}" = "dns-01" ]] && [[ -z "${HOOK}" ]]; then
     _exiterr "Challenge type dns-01 needs a hook script for deployment... cannot continue."
   fi
@@ -516,13 +516,21 @@
 
   set +e
   if [[ "${1}" = "head" ]]; then
+      echo "HTTP_REQUEST: HEAD -> '${2}'" > /dev/stderr
+      echo "" > /dev/stderr
     statuscode="$(curl ${ip_version:-} ${CURL_OPTS} -A "dehydrated/${VERSION} curl/${CURL_VERSION}" -s -w "%{http_code}" -o "${tempcont}" "${2}" -I)"
     touch "${tempheaders}"
     curlret="${?}"
   elif [[ "${1}" = "get" ]]; then
+    echo "HTTP_REQUEST: GET -> '${2}'" > /dev/stderr
+    echo "" > /dev/stderr
+    echo "-- end get" > /dev/stderr
     statuscode="$(curl ${ip_version:-} ${CURL_OPTS} -A "dehydrated/${VERSION} curl/${CURL_VERSION}" -L -s -w "%{http_code}" -o "${tempcont}" -D "${tempheaders}" "${2}")"
     curlret="${?}"
   elif [[ "${1}" = "post" ]]; then
+    echo "HTTP_REQUEST: POST -> '${2}'" > /dev/stderr
+      echo "" > /dev/stderr
+      echo "-- end post" > /dev/stderr
     statuscode="$(curl ${ip_version:-} ${CURL_OPTS} -A "dehydrated/${VERSION} curl/${CURL_VERSION}" -s -w "%{http_code}" -o "${tempcont}" "${2}" -D "${tempheaders}" -H 'Content-Type: application/jose+json' -d "${3}")"
     curlret="${?}"
   else
@@ -531,6 +539,15 @@
   fi
   set -e
 
+  echo "HTTP HEADERS:" > /dev/stderr
+  cat ${tempheaders} > /dev/stderr
+  echo "" > /dev/stderr
+  echo "HTTP RESPONSE:" > /dev/stderr
+  cat ${tempcont} > /dev/stderr
+  echo "" > /dev/stderr
+  echo "-- end response" > /dev/stderr
+  echo "" > /dev/stderr
+
   if [[ ! "${curlret}" = "0" ]]; then
     _exiterr "Problem connecting to server (${1} for ${2}; curl returned with ${curlret})"
   fi
@@ -747,10 +764,14 @@
     # Prepare challenge tokens and deployment parameters
     keyauth="${challenge_tokens[${idx}]}.${thumbprint}"
 
+    echo "FREDRIK: CHALLENGETYPE ${CHALLENGETYPE}"
     case "${CHALLENGETYPE}" in
       "http-01")
         # Store challenge response in well-known location and make world-readable (so that a webserver can access it)
-        printf '%s' "${keyauth}" > "${WELLKNOWN}/${challenge_tokens[${idx}]}"
+          printf '%s' "${keyauth}" > "${WELLKNOWN}/${challenge_tokens[${idx}]}"
+	  echo "FREDRIK: FILE ${WELLKNOWN}/${challenge_tokens[${idx}]}:"
+	  cat "${WELLKNOWN}/${challenge_tokens[${idx}]}"
+	  echo "EOF"
         chmod a+r "${WELLKNOWN}/${challenge_tokens[${idx}]}"
         keyauth_hook="${keyauth}"
         ;;
@@ -762,6 +783,10 @@
         keyauth_hook="$(printf '%s' "${keyauth}" | "${OPENSSL}" dgst -sha256 -c -hex | awk '{print $2}')"
         generate_alpn_certificate "${identifier}" "${keyauth_hook}"
         ;;
+      "x-sunet-01")
+          keyauth_hook="${keyauth}"
+	  echo "X-SUNET-01: KEYAUTH ${keyauth}"
+	  ;;
     esac
 
     keyauths[${idx}]="${keyauth}"
@@ -848,13 +873,28 @@
   fi
 
   # Finally request certificate from the acme-server and store it in cert-${timestamp}.pem and link from cert.pem
-  echo " + Requesting certificate..."
   csr64="$( <<<"${csr}" "${OPENSSL}" req -config "${OPENSSL_CNF}" -outform DER | urlbase64)"
   if [[ ${API} -eq 1 ]]; then
+    echo " + Requesting certificate..."
     crt64="$(signed_request "${CA_NEW_CERT}" '{"resource": "new-cert", "csr": "'"${csr64}"'"}' | "${OPENSSL}" base64 -e)"
     crt="$( printf -- '-----BEGIN CERTIFICATE-----\n%s\n-----END CERTIFICATE-----\n' "${crt64}" )"
   else
-    result="$(signed_request "${finalize}" '{"csr": "'"${csr64}"'"}' | clean_json | get_json_string_value certificate)"
+    order="$(echo ${finalize} | sed 's,/*[^/]\+/*$,,')"
+    result="$(signed_request "${order}" '{}' | clean_json)"
+    reqstatus="$(printf '%s\n' "${result}" | get_json_string_value status)"
+    if [[ "${reqstatus}" = "ready" ]]; then
+      echo " + Finalizing order..."
+      result="$(signed_request "${finalize}" '{"csr": "'"${csr64}"'"}' | clean_json)"
+      reqstatus="$(printf '%s\n' "${result}" | get_json_string_value status)"
+    fi
+    echo " + Requesting certificate..."
+    while [[ "${reqstatus}" = "processing" ]]; do
+      sleep 1
+      result="$(signed_request "${order}" '{}' | clean_json)"
+      reqstatus="$(printf '%s\n' "${result}" | get_json_string_value status)"
+      echo " + Waiting for certificate processing..."
+    done
+    result="$(printf "${result}" | get_json_string_value certificate)"
     crt="$(signed_request "${result}" "")"
   fi
 
